main.py

```diff
@@ -25,12 +25,41 @@ def create_app(application):
 def publish_govuk_alerts(self):
     from flask import current_app
 
-    alerts = alerts_from_yaml()
-    rendered_pages = get_rendered_pages(alerts)
+    try:
+        alerts = alerts_from_yaml()
+        rendered_pages = get_rendered_pages(alerts)
 
-    upload_to_s3(current_app.config, rendered_pages)
+        upload_to_s3(current_app.config, rendered_pages)
 
-    purge_cache(current_app.config)
+        purge_cache(current_app.config)
+    except AlertsFromYamlError:
+        current_app.logger.exception("Failed to get the alerts from yaml: " + AlertsFromYamlError)
+        self.retry(queue=current_app.config['QUEUE_NAME'])
+        return
+
+    except RenderPagesError:
+        current_app.logger.exception("Failed to render the pages: " + RenderPagesError)
+        self.retry(queue=current_app.config['QUEUE_NAME'])
+        return
+
+    except S3UploadError:
+        current_app.logger.exception("Failed to upload to S3: " + S3UploadError)
+        self.retry(queue=current_app.config['QUEUE_NAME'])
+        return
+
+    except PurgeCacheError:
+        current_app.logger.exception("Failed to upload to S3: " + PurgeCacheError)
+        self.retry(queue=current_app.config['QUEUE_NAME'])
+        return
+
+    except self.MaxRetriesExceededError:
+        current_app.logger.exception("Max retry failed: Failed to publish content to gov.uk/alerts")
+        return
+
+    except Exception:
```

Comment:
Could we just have this and skip all the other custom exceptions? The other exceptions are just rescuing `Exception` in any case, and it should be clear from the exception message and stack trace where the error occurred.

(Sorry, I thought I wrote this comment already, but it seems to have disappeared.)

Reply:
I think it'll be informative enough if we just rescue Exception. A few advantages to this approach:

- It's less code.
- We can always add more precise exception handling in future, when we'll know more about what we need.
- It's easier to test, since we only have one handler (versus handlers in each function and at the top-level).

Reply:
Done here: https://github.com/alphagov/notifications-govuk-alerts/pull/151/commits/f868b0d00f980f28867611b1c3802d9a4810ef88